src/ballet/shred/test_deshredder.c

```diff
@@ -0,0 +1,135 @@
+
+    size_t n = fread( shred_buf, 1, (ulong)hdr.filesz, file );
+    FD_TEST(( n==(ulong)hdr.filesz ));
+
+    /* Parse shred */
+    fd_shred_t const * shred = fd_shred_parse( shred_buf, (ulong)hdr.filesz );
+
+    /* Refill deshredder with shred */
+    fd_shred_t const * const shred_list[1] = { shred };
+    deshred.shreds    = shred_list;
+    deshred.shred_cnt = 1U;
+
+    fd_deshredder_next( &deshred );
+  }
+
+  /* Did we gracefully finish consuming the archive? */
+  FD_TEST( ar_err==ENOENT );
+
+  /* Check size of defragmented batch */
+  ulong batch_sz = sizeof(batch) - deshred.bufsz;
+  /* fwrite( batch, batch_sz, 1, stdout ); */
+  FD_TEST( batch_sz==3080UL                );
+  FD_TEST( batch_sz==localnet_batch_0_0_sz );
+
+  /* Check number of shreds */
+  ulong shred_cnt = *(ulong *)batch;
+  FD_TEST( shred_cnt==64UL );
+
+  /* Verify deshredded content */
+  FD_TEST( 0==memcmp( batch, localnet_batch_0_0, localnet_batch_0_0_sz ) );
+
+  FD_LOG_NOTICE(( "v14 slot 0" ));
+  file = fmemopen( (void *)localnet_v14_shreds_0, localnet_v14_shreds_0_sz, "r" );
+  FD_TEST( file );
+  FD_TEST( 0==fd_ar_read_init( file ) );
+
+  fd_deshredder_init( &deshred, &batch, sizeof(batch), NULL, 0 );
+
+  while( (ar_err = fd_ar_read_next( file, &hdr ))==0 ) {
+    uchar shred_buf[ FD_SHRED_SZ ];
+
+    /* Read next file from archive */
+    FD_TEST(( hdr.filesz>=0 && hdr.filesz < (long)sizeof(shred_buf) ));
+
+    size_t n = fread( shred_buf, 1, (ulong)hdr.filesz, file );
+    FD_TEST(( n==(ulong)hdr.filesz ));
+
+    /* Parse shred */
+    fd_shred_t const * shred = fd_shred_parse( shred_buf, (ulong)hdr.filesz );
+
+    /* Refill deshredder with shred */
+    fd_shred_t const * const shred_list[1] = { shred };
+    deshred.shreds    = shred_list;
+    deshred.shred_cnt = 1U;
+
+    fd_deshredder_next( &deshred );
+  }
+
+  /* Did we gracefully finish consuming the archive? */
+  FD_TEST( ar_err==ENOENT );
+
+  FD_LOG_NOTICE(( "v14 slot 1" ));
+  file = fmemopen( (void *)localnet_v14_shreds_1, localnet_v14_shreds_1_sz, "r" );
+  FD_TEST( file );
+  FD_TEST( 0==fd_ar_read_init( file ) );
+
+  fd_deshredder_init( &deshred, &batch, sizeof(batch), NULL, 0 );
+
+  while( (ar_err = fd_ar_read_next( file, &hdr ))==0 ) {
+    uchar shred_buf[ FD_SHRED_SZ ];
+
+    /* Read next file from archive */
+    FD_TEST(( hdr.filesz>=0 && hdr.filesz < (long)sizeof(shred_buf) ));
+
+    size_t n = fread( shred_buf, 1, (ulong)hdr.filesz, file );
+    FD_TEST(( n==(ulong)hdr.filesz ));
+
+    /* Parse shred */
+    fd_shred_t const * shred = fd_shred_parse( shred_buf, (ulong)hdr.filesz );
+    FD_TEST( shred );
+
+    /* Refill deshredder with shred */
+    fd_shred_t const * const shred_list[1] = { shred };
+    deshred.shreds    = shred_list;
+    deshred.shred_cnt = 1U;
+
+    fd_deshredder_next( &deshred );
+  }
+
+  /* Did we gracefully finish consuming the archive? */
+  FD_TEST( ar_err==ENOENT );
```

Comment:
This could be 4 calls to a test function instead of mostly duplicated chunks, no?

Reply:
Can you add a test with chained?